service/controller/resource/subnet/resource.go

```diff
@@ -0,0 +1,273 @@
+				Parameters: key.ToParameters(parameters),
+				Template:   armTemplate,
+			},
+		}
+
+		// We only submit the deployment if it doesn't exist or it exists but it's out of date.
+		shouldSubmitDeployment := currentDeployment.IsHTTPStatus(404)
+		if !shouldSubmitDeployment {
+			shouldSubmitDeployment, err = r.isDeploymentOutOfDate(ctx, cr, currentDeployment)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+
+		if shouldSubmitDeployment {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "template or parameters changed")
+			err = r.createDeployment(ctx, deploymentsClient, key.ClusterID(&cr), deploymentName, desiredDeployment)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
+			reconciliationcanceledcontext.SetCanceled(ctx)
+
+			return nil
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "template and parameters unchanged")
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment is in state '%s'", *currentDeployment.Properties.ProvisioningState))
+
+		if key.IsFailedProvisioningState(*currentDeployment.Properties.ProvisioningState) {
+			r.debugger.LogFailedDeployment(ctx, currentDeployment, err)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "removing failed deployment")
+			_, err = deploymentsClient.Delete(ctx, key.ClusterID(&cr), deploymentName)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) isDeploymentOutOfDate(ctx context.Context, cr capzv1alpha3.AzureCluster, currentDeployment azureresource.DeploymentExtended) (bool, error) {
+	crVersion := strconv.FormatInt(cr.ObjectMeta.Generation, 10)
+	currentParams, ok := currentDeployment.Properties.Parameters.(map[string]interface{})
+	if !ok {
+		return false, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", map[string]interface{}{}, currentDeployment.Properties.Parameters)
+	}
+
+	deploymentVersion, ok := currentParams["azureClusterVersion"].(map[string]interface{})["value"].(string)
+	if !ok {
+		return false, microerror.Maskf(wrongTypeError, "expected 'string', got '%T'", currentDeployment.Properties.Parameters)
+	}
+
+	r.logger.LogCtx(ctx, "message", "Checking if deployment is out of date", "azureClusterVersion", crVersion, "deploymentParameter", currentParams["azureClusterVersion"])
+
+	return crVersion != deploymentVersion, nil
+}
+
+func (r *Resource) getDeploymentParameters(ctx context.Context, clusterID, azureClusterVersion, virtualNetworkName string, allocatedSubnet *capzv1alpha3.SubnetSpec) (map[string]interface{}, error) {
+	return map[string]interface{}{
+		"azureClusterVersion": azureClusterVersion,
+		"nateGatewayName":     "workers-nat-gw",
+		"nodepoolName":        allocatedSubnet.Name,
+		"routeTableName":      fmt.Sprintf("%s-%s", clusterID, "RouteTable"),
+		"securityGroupName":   fmt.Sprintf("%s-%s", clusterID, "WorkerSecurityGroup"),
```

Comment:
I'm assuming here that the route table and security group will have this name. Do you think it will be better to fetch them from the ARM Deployment outputs? It will add Azure API calls, but this code will be independent from the naming.

Reply:
What about having functions in `key` package for getting NAT gateway, route table and security group name?

Reply:
We could do that. I would need to change the `deployment` handler as well, because the naming is hardcoded in the ARM template.

Reply:
It would be ideal to have the names in one place. Not sure how we did it before for `masters` and `instance` resource. Not a blocker this one :)